### PR TITLE
[Build] Define math constants for MSVC targets

### LINKED ISSUE #54459: [Bug] [Portability][MSVC]: M_LOG2E is unavailable when building Flash Attention with NVCC and MSVC
### Environment

- Windows Server 2025
- Visual Studio Build Tools with MSVC
- CUDA/NVCC 13.2
- Python 3.12
- PyTorch 2.11.0+cu130
- SystemPanic/vllm-windows 0.25-based deployment
- DeepSeek-V4-Flash-0731
- 4 x H200 NVL GPUs, SM90a

### Background

I encountered this issue while building vllm-project/flash-attention through SystemPanic/vllm-windows for a deployment serving DeepSeek-V4-Flash-0731 on 4 x H200 NVL GPUs.

### Problem

Several Flash Attention CUDA source files use `M_LOG2E` in softmax and `exp2f` scale calculations:

    csrc/flash_attn/flash_api.cpp
    csrc/flash_attn/src/softmax.h
    hopper/flash_bwd_preprocess_kernel.h
    hopper/mainloop_fwd_sm80.hpp
    hopper/mainloop_fwd_sm90_tma_gmma_ws.hpp
    hopper/mainloop_bwd_sm80.hpp
    hopper/mainloop_bwd_sm90_tma_gmma_ws.hpp
    hopper/softmax.h

Examples include:

    params.scale_softmax_log2 =
        softmax_scale * M_LOG2E;

and:

    const float max_scaled =
        max(mi) == -INFINITY
            ? 0.f
            : max(mi) * (
                  Scale_max ? scale : float(M_LOG2E));

`M_LOG2E` represents `log2(e)` and is commonly available in GNU/Linux build environments. Its availability is implementation-specific, however, and MSVC normally exposes the `M_*` constants only when `_USE_MATH_DEFINES` is defined before the relevant math header is included.

I understand that native Windows may not be part of the supported build matrix. This appears to be a small source-portability issue rather than a request for broader Windows support.

### Tested workaround

The working downstream build replaces `M_LOG2E` with the corresponding float constant:

    - float(M_LOG2E)
    + 1.4426950408889634f

For example:

    - params.scale_softmax_log2 =
    -     softmax_scale * M_LOG2E;
    + params.scale_softmax_log2 =
    +     softmax_scale * 1.4426950408889634f;

The replacement preserves the intended `log2(e)` scale used by the float-based softmax and `exp2f` paths.

Applying the replacement to all affected occurrences allowed the Flash Attention extensions to build successfully with NVCC and MSVC.

### Related PyTorch-side solution

The discussion in: https://github.com/vllm-project/vllm/pull/54007

indicates that a PyTorch-side update may resolve this compatibility issue in the Windows build environment and perhaps could the same change be done here as well.

### Other possible solutions

The affected NVCC and MSVC targets could define:

    _USE_MATH_DEFINES=1

For example:

    target_compile_definitions(
        <target> PRIVATE _USE_MATH_DEFINES=1
    )

Alternatively, Flash Attention could use a project-owned `constexpr` value for `log2(e)` instead of depending on the host math-header configuration.

The source-level constant would be independent of PyTorch headers, build flags, and math-header include order.

### Repository context

Seven affected files are inherited from Dao-AILab/flash-attention.

The vllm-project fork additionally contains an affected occurrence in:

    hopper/softmax.h

I am reporting the complete set here because this is the Flash Attention fork built by vLLM.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by Blade-of-Mikasa**
## Summary

- define `_USE_MATH_DEFINES` for the FA2 and FA3 extension targets when building with MSVC
- keep the definition `PRIVATE` to those targets, leaving non-MSVC builds unchanged

Fixes vllm-project/vllm#54459.

## Why this approach

MSVC does not expose the non-standard `M_*` constants from `<cmath>` unless `_USE_MATH_DEFINES` is defined before the header is included. This is the supported opt-in described by Microsoft:
https://learn.microsoft.com/en-us/cpp/c-runtime-library/math-constants?view=msvc-170

There are several existing precedents for handling the same portability issue at the build layer:

- this repository's `hopper/setup.py` already passes `-D_USE_MATH_DEFINES` for `win_amd64` NVCC builds:
  https://github.com/vllm-project/flash-attention/blob/main/hopper/setup.py#L595-L600
- upstream Dao-AILab FlashAttention uses the same Windows NVCC definition:
  https://github.com/Dao-AILab/flash-attention/blob/main/hopper/setup.py
- DeepSeek FlashMLA also passes `-D_USE_MATH_DEFINES` in its CUDA extension build:
  https://github.com/deepseek-ai/FlashMLA/blob/main/setup.py
- the analogous vLLM FlashMLA fix uses MSVC-guarded, target-scoped compile definitions:
  https://github.com/vllm-project/vllm/pull/54007

Using `VLLM_FA_GPU_FLAGS` alone would not be sufficient: `define_gpu_extension_target` applies those flags only to the GPU language, while `M_LOG2E` is also used by host C++ sources such as `csrc/flash_attn/flash_api.cpp`. Target compile definitions cover both the C++ and CUDA translation units.

## Duplicate check

Searched open pull requests in `vllm-project/flash-attention` for `M_LOG2E`, `_USE_MATH_DEFINES`, `MSVC`, and `54459`; no active duplicate was found.

## Testing

- `git diff --check`
- Windows MSVC smoke test:
  https://github.com/Blade-of-Mikasa/flash-attention/actions/runs/33410272825
  - without `_USE_MATH_DEFINES`: fails as expected with `C2065: 'M_LOG2E': undeclared identifier`
  - with `_USE_MATH_DEFINES`: compiles successfully
  - workflow result: success with no annotations
- not run: a full Windows CUDA/NVCC build, because that environment is not available locally
- model evaluation: not applicable; this is a build-only portability change

## AI assistance

OpenAI Codex assisted with issue investigation, the CMake change, the MSVC smoke test, and drafting this description. I reviewed and understand every changed line and the stated validation boundary.

### CODE REVIEW

**Comment by Blade-of-Mikasa** [comment]
Hi @LucasWilkinson, would you mind taking a look when you have a chance? This is a small MSVC build portability fix following the approach in vllm-project/vllm#54007. Thanks!
